### ISSUE TITLE
Evaluations should work for runtimeType

### ISSUE THREAD

**Issue by FMorschel** [dart-lang/sdk#59711]
Say you have

```dart
class A {}

class B extends A {
  int foo() => 0;
}

class C {
  C(this.a);
  A a;
}

void main() {
  var c = C(B());  // Breakpoint here 
}
```

Now you try to evaluate `c.a.foo()` it gives you an error message about `foo` not being defined for `A`. But if you do `(c.a as dynamic).foo()` everything works normally and it can result in `0` here.

Could the evaluations do something like this internally?

// CC @DanTup

**Comment by mraleph** [comment]
Right now evaluated expression are compiled as Dart code. So you can only do things which are expressible in Dart. 

We used to compile expressions without preserving static types (e.g. `c` would be `dynamic` and consequently `c.a` would be `dynamic` as well), but that hits a problem that any features which rely on static type and type inference (e.g. extensions and extension types or generics) don't work as one would expect.

Making `c.a.foo` work reliably across all possibilities requires some sophisticated machinery because it requires parser to somehow be able to introspect runtime state.

**Comment by a-siva** [comment]
I am not sure if it is a good idea for the VM or CFE to evaluate the expression differently from how it would behave when the code is normally run.

**Comment by FMorschel** [comment]
The problem currently is that the expression is not running normally, but actually giving out an error. While on code it does run just fine.

**Comment by a-siva** [comment]
I modified your code as follows
```
class A {}

class B extends A {
  int foo() => 0;
}

class C {
  C(this.a);
  A a;
}

void main() {
  var c = C(B());  // Breakpoint here 
  print(c.a.foo());
}
```

and trying running it, this is what I get
```

../../../tmp/test.dart:14:13: Error: The method 'foo' isn't defined for the class 'A'.
 - 'A' is from '../../../tmp/test.dart'.
Try correcting the name to the name of an existing method, or defining a method named 'foo'.
  print(c.a.foo());
```

**Comment by FMorschel** [comment]
Yeah. The code where I saw the example was inside one of the files for the analyzer and it was tested for the type with if case.

It also shows different getters in the evaluation hover and such. That is why I thought it should work.

**Comment by a-siva** [comment]
> Yeah. The code where I saw the example was inside one of the files for the analyzer and it was tested for the type with if case.
> 
> It also shows different getters in the evaluation hover and such. That is why I thought it should work.

The expression evaluator does not have the context of the type testing that the code you were running has.

**Comment by DanTup** [comment]
I'm actually confused... The original code where @FMorschel saw this was like this:

![image](https://github.com/user-attachments/assets/0f77fe7d-5b2f-4c2e-8fbf-67b183c285b8)

And the evaluation failed like this:

![image](https://github.com/user-attachments/assets/4dcd6682-e6e8-4cb9-8fac-421aba4e5865)

However, in a simple test of what I think is the same, I actually don't get an error - I am allowed to evaluate `a.x` even though the static type of `a` does not have an `x`:

```dart
void main() {
  f(B('test'));
}

void f(A a) {
  print(a);
}

class A {}

class B extends A {
  final String x;
  B(this.x);
}
```

![image](https://github.com/user-attachments/assets/ca433b50-24ae-41dd-9299-46599a8a0432)

I don't understand why these are different.


@a-siva 

> I am not sure if it is a good idea for the VM or CFE to evaluate the expression differently from how it would behave when the code is normally run.

I can understand that, but I also feel like having to add `?`/`!` and casts to an expression just for quick evaluations while debugging can be quite annoying - and if you're hovering over an expression in the file, it might not be possible to amend it anyway. However, if it prevents using things like extension methods, maybe it's the right trade.

**Comment by FMorschel** [comment]
@a-siva, I guess what @DanTup was trying to say was that:

For this code (similar to his above):

```dart
void main() {
  f(B(B()));
}

void f(A a) {
  print(a);
  a.x;
//  ^ undefined_getter
}

class A {}

class B extends A {
  final A? x;
  B([this.x]);
}
```

We can't run it because it doesn't compile. But if you remove the `a.x` line and evaluate `a.x` in the debugger at the `print(a)` line, it works just fine.

From what I could gather, this "promotion" of sorts happens for "known variables", so this feels a bit inconsistent for me. See this (I've added lines 7 and 8 after breaking, so it is fine):

<img width="585" height="185" alt="Image" src="https://github.com/user-attachments/assets/024aa4c0-a181-45c7-ba56-992eb190864e" />

We could even evaluate `a.x.x` there, even though `x` is `A?`, so there is definitely some promotion happening for more levels, I'm just not entirely sure what the rule is.

---

**Edit**

Even getters are promoted, see:

<img width="628" height="566" alt="Image" src="https://github.com/user-attachments/assets/fb18651f-e455-49e0-a34b-29571b92a1a4" />

<details>

<summary> Code </summary>

```dart
void main() {
  f(B(C(C())));
}

void f(A a) {
  print(a);
}

class A {}

class B extends A {
  final A? x;
  B([this.x]);
}

class C extends B {
  C([super.x]);
  A? get y => x;
}
```

</details>
